### PR TITLE
Implements workaround for C* bug (duplicates in list on insert/update…

### LINKED ISSUE #138: Duplicate values observed in List cells when running Migrate or Validation multiple times
Duplicate values observed in List cells when running Migrate or Validation multiple times.

**Note:** This issue is a [bug in Cassandra](https://issues.apache.org/jira/browse/CASSANDRA-11368) (List inserts are not truly idempotent) & not really a CDM issue.

**Details of the issue:** As a side effect of the above C* bug, when CDM Migration or Validation (with auto-correct) is executed multiple times on the same table (this is not common, but can happen for various reasons), any existing list fields on target are overwritten again with the same timestamp (list values & timestamp from source), and this instead of replacing the full list duplicates all the entries on list. If you run the tool N times, the values would be duplicated N times.

**Priority/Customer Impact:** IHAC who is in midst of migrating several DSE 5.1 clusters to DSE 6.8. They are using ZDM + CDM for the same & are facing this issue. They understand that this is not a CDM issue, however they have asked us to provide a possible workaround/fix.

**Workaround/Fix:** The issue (duplicate cell values in List) happens only the timestamp is exactly same. If we increment the long epoch timestamp value by 1 (_which is 1 microseconds (1/1,000,000 second)_) & use this as the timestamp, the issue goes away. The timestamp value also is almost unchanged for humans and it does not create any (known) issues with migration (ZDM and/or CDM). Hence we will conditionally allow users to add an arbitrary long value (this should be typically 1) to the timestamp, which will allow them to workaround this issue.
Screenshot 2023-05-02 at 9 53 23 AM

### PR DESCRIPTION
**Pull request by pravinbhat**
**What this PR does**:
Implements workaround for C* bug (duplicates in list on insert/update with same timestamp) https://issues.apache.org/jira/browse/CASSANDRA-11368

**Which issue(s) this PR fixes**:
Fixes #138 

**Checklist:**
- [ ] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 1]
```diff
@@ -30,6 +30,8 @@ public class AbstractJobSession extends BaseJobSession {
     protected List<String> ttlWTCols;
     protected String tsReplaceValStr;
     protected long tsReplaceVal;
+    protected Long customWritetime = 0l;
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:33]:
```suggestion
    protected long customWriteTime = 0l;
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 2]
```diff
@@ -30,6 +30,8 @@ public class AbstractJobSession extends BaseJobSession {
...
+    protected Long incrWritetime = 0l;
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:34]:
```suggestion
    protected long incrWriteTime = 0l;
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 3]
```diff
@@ -95,13 +96,21 @@ protected AbstractJobSession(CqlSession sourceSession, CqlSession astraSession,
...
-        if (null != customWriteTimeStr && customWriteTimeStr.trim().length() > 1 && StringUtils.isNumeric(customWriteTimeStr.trim())) {
+                Util.getSparkPropOr(sc, "spark.target.writeTime.fixedValue", "0");
+        if (null != customWriteTimeStr && customWriteTimeStr.trim().length() > 0 && StringUtils.isNumeric(customWriteTimeStr.trim())) {
             customWritetime = Long.parseLong(customWriteTimeStr);
         }
 
+        String incrWriteTimeStr =
+                Util.getSparkPropOr(sc, "spark.target.writeTime.incrementBy", "0");
+        if (null != incrWriteTimeStr && incrWriteTimeStr.trim().length() > 0 && StringUtils.isNumeric(incrWriteTimeStr.trim())) {
+            incrWritetime = Long.parseLong(incrWriteTimeStr);
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:107]:
```suggestion
            incrWriteTime = Long.parseLong(incrWriteTimeStr);
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 4]
```diff
@@ -95,13 +96,21 @@ protected AbstractJobSession(CqlSession sourceSession, CqlSession astraSession,
...
+        }
+
         logger.info("PARAM -- TTL-WriteTime Columns: {}", ttlWTCols);
-        logger.info("PARAM -- WriteTimestampFilter: {}", writeTimeStampFilter);
+        logger.info("PARAM -- WriteTimes Filter: {}", writeTimeStampFilter);
+        logger.info("PARAM -- WriteTime Custom Value: {}", customWritetime);
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:112]:
```suggestion
        logger.info("PARAM -- WriteTime Custom Value: {}", customWriteTime);
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 5]
```diff
@@ -95,13 +96,21 @@ protected AbstractJobSession(CqlSession sourceSession, CqlSession astraSession,
...
+        logger.info("PARAM -- WriteTime increment Value: {}", incrWritetime);
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:113]:
```suggestion
        logger.info("PARAM -- WriteTime increment Value: {}", incrWriteTime);
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 6]
```diff
@@ -196,7 +205,7 @@ public BoundStatement bindInsert(PreparedStatement insertStatement, Row sourceRo
                 if (customWritetime > 0) {
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:205]:
```suggestion
                if (customWriteTime > 0) {
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 7]
```diff
@@ -196,7 +205,7 @@ public BoundStatement bindInsert(PreparedStatement insertStatement, Row sourceRo
...
                     boundInsertStatement = boundInsertStatement.set(index, customWritetime, Long.class);
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:206]:
```suggestion
                    boundInsertStatement = boundInsertStatement.setLong(index, customWriteTime);
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 8]
```diff
@@ -196,7 +205,7 @@ public BoundStatement bindInsert(PreparedStatement insertStatement, Row sourceRo
...
                 } else {
-                    boundInsertStatement = boundInsertStatement.set(index, getLargestWriteTimeStamp(sourceRow), Long.class);
+                    boundInsertStatement = boundInsertStatement.set(index, getLargestWriteTimeStamp(sourceRow) + incrWritetime, Long.class);
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:208]:
```suggestion
                    boundInsertStatement = boundInsertStatement.setLong(index, getLargestWriteTimeStamp(sourceRow) + incrWriteTime);
```

**Review by msmygit — APPROVED** [review]
LGTM. I left trivial nitpick comments.

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pravinbhat) [reply src/main/java/datastax/astra/migrate/AbstractJobSession.java:33]: Good observation, should be primitive, will fix.

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 9]
```diff
@@ -95,13 +96,21 @@ protected AbstractJobSession(CqlSession sourceSession, CqlSession astraSession,
         }
 
         String customWriteTimeStr =
-                Util.getSparkPropOr(sc, "spark.target.custom.writeTime", "0");
-        if (null != customWriteTimeStr && customWriteTimeStr.trim().length() > 1 && StringUtils.isNumeric(customWriteTimeStr.trim())) {
-            customWritetime = Long.parseLong(customWriteTimeStr);
+                Util.getSparkPropOr(sc, "spark.target.writeTime.fixedValue", "0");
+        if (null != customWriteTimeStr && customWriteTimeStr.trim().length() > 0 && StringUtils.isNumeric(customWriteTimeStr.trim())) {
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:100]: Could we please use the lang3 apache commons here to further simply it as,
```suggestion
        if (StringUtils.isNotBlank(customWriteTimeStr) && StringUtils.isNumeric(customWriteTimeStr.trim())) {
```

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 10]
```diff
@@ -95,13 +96,21 @@ protected AbstractJobSession(CqlSession sourceSession, CqlSession astraSession,
...
+            customWriteTime = Long.parseLong(customWriteTimeStr);
+        }
+
+        String incrWriteTimeStr =
+                Util.getSparkPropOr(sc, "spark.target.writeTime.incrementBy", "0");
+        if (null != incrWriteTimeStr && incrWriteTimeStr.trim().length() > 0 && StringUtils.isNumeric(incrWriteTimeStr.trim())) {
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:106]:
```suggestion
        if (StringUtils.isNotBlank(incrWriteTimeStr) && StringUtils.isNumeric(incrWriteTimeStr.trim())) {
```

**Review by msmygit — APPROVED** [review]
👍🏼

File: src/main/java/datastax/astra/migrate/AbstractJobSession.java [hunk 11]
```diff
@@ -90,18 +91,26 @@ protected AbstractJobSession(CqlSession sourceSession, CqlSession astraSession,
...
-        if (null != customWriteTimeStr && customWriteTimeStr.trim().length() > 1 && StringUtils.isNumeric(customWriteTimeStr.trim())) {
-            customWritetime = Long.parseLong(customWriteTimeStr);
+                Util.getSparkPropOr(sc, "spark.target.writeTime.fixedValue", "0");
+        if (StringUtils.isNotBlank(customWriteTimeStr) && StringUtils.isNumeric(customWriteTimeStr.trim())) {
+            customWriteTime = Long.parseLong(customWriteTimeStr);
+        }
+
+        String incrWriteTimeStr =
+                Util.getSparkPropOr(sc, "spark.target.writeTime.incrementBy", "0");
+        if (StringUtils.isNotBlank(incrWriteTimeStr) && StringUtils.isNumeric(incrWriteTimeStr.trim())) {
```
Comment (msmygit) [inline src/main/java/datastax/astra/migrate/AbstractJobSession.java:106]:
```suggestion
        if (StringUtils.isNotBlank(incrWriteTimeStr) && StringUtils.isNumeric(incrWriteTimeStr)) {
```

**Pull request merged**